src/db.c

```diff
@@ -463,41 +463,40 @@ long long emptyDb(int dbnum, int flags, void(callback)(dict*)) {
 }
 
 /* Initialize temporary db on replica for use during diskless replication. */
-tempDb *initTempDb(void) {
-    tempDb *tempDb = zmalloc(sizeof(tempDb));
-
-    tempDb->dbarray = zcalloc(sizeof(redisDb)*server.dbnum);
+redisDb *initTempDb(void) {
+    redisDb *tempDb = zcalloc(sizeof(redisDb)*server.dbnum);
     for (int i=0; i<server.dbnum; i++) {
-        tempDb->dbarray[i].dict = dictCreate(&dbDictType);
-        tempDb->dbarray[i].expires = dictCreate(&dbExpiresDictType);
+        tempDb[i].dict = dictCreate(&dbDictType);
+        tempDb[i].expires = dictCreate(&dbExpiresDictType);
+        tempDb[i].slots_to_keys = NULL;
     }
 
     if (server.cluster_enabled) {
-        /* Prepare temp slots_to_keys to be written during async diskless replication. */
-        slotToKeyTempDbFlush();
+        /* Prepare temp slots to keys map to be written during async diskless replication. */
+        tempDb->slots_to_keys = zmalloc(sizeof(*tempDb->slots_to_keys));
+        slotToKeyFlush(tempDb);
     }
 
     return tempDb;
 }
 
 /* Discard tempDb, this can be slow (similar to FLUSHALL), but it's always async */
- void discardTempDb(tempDb *tempDb, void(callback)(dict*)) {
+ void discardTempDb(redisDb *tempDb, void(callback)(dict*)) {
     int async = 1;
 
     /* Release temp DBs */
-    emptyDbStructure(tempDb->dbarray, -1, async, callback);
+    emptyDbStructure(tempDb, -1, async, callback);
     for (int i=0; i<server.dbnum; i++) {
-        dictRelease(tempDb->dbarray[i].dict);
-        dictRelease(tempDb->dbarray[i].expires);
+        dictRelease(tempDb[i].dict);
+        dictRelease(tempDb[i].expires);
     }
 
-    zfree(tempDb->dbarray);
-    zfree(tempDb);
-
     if (server.cluster_enabled) {
-        /* Flush temp slots_to_keys. */
-        slotToKeyTempDbFlush();
+        /* Flush temp slots to keys map. */
+        slotToKeyFlush(tempDb);
     }
+
+    zfree(tempDb);
 }
```

Comment:
The slot-to-key array needs to be free'd somewhere before the tempDb is free'd. Flush just memsets it to zero.

Maybe we can replace slotToKeyFlush with two functions: slotToKeyNew (allocates it) and slotToKeyFree (frees it)?